### PR TITLE
Make ExpressionPromoter plugable

### PR DESCRIPTION
**Pull request by david-garcia-garcia**
Now that the library allows to instantiate more .Net types easily, Expression promoter needs to be more powerful depending on the needs of the destination types.

This PR makes it plugable and replaceable from outside the library.

### CODE REVIEW

File: src/System.Linq.Dynamic.Core/CustomTypeProviders/IDynamicLinkCustomTypeProvider.cs [hunk 1]
```diff
@@ -20,5 +20,6 @@ public interface IDynamicLinkCustomTypeProvider
         /// <param name="typeName">The typename to resolve.</param>
         /// <returns>A resolved <see cref="Type"/> or null when not found.</returns>
         Type ResolveType([NotNull] string typeName);
+
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/CustomTypeProviders/IDynamicLinkCustomTypeProvider.cs]: please remove this empty line.

File: src/System.Linq.Dynamic.Core/Parser/ExpressionPromoter.cs [hunk 2]
```diff
@@ -3,9 +3,9 @@
 
 namespace System.Linq.Dynamic.Core.Parser
 {
-    internal static class ExpressionPromoter
+    internal class ExpressionPromoter : IExpressionPromoter
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionPromoter.cs]:
```suggestion
    maybe if you have time ; can you create a simple test which tests this ExpressionPromoter?
```

File: src/System.Linq.Dynamic.Core/Parser/ExpressionPromoter.cs [hunk 3]
```diff
@@ -3,9 +3,9 @@
...
     {
-        public static Expression Promote(Expression expr, Type type, bool exact, bool convertExpr)
+        public Expression Promote(Expression expr, Type type, bool exact, bool convertExpr)
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/ExpressionPromoter.cs]: Add `/// <inheritdoc cref="IExpressionPromoter.Promote(Expression, Type, bool, bool)"/>` here.

File: src/System.Linq.Dynamic.Core/Parser/IExpressionPromoter.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Linq.Expressions;
+
+namespace System.Linq.Dynamic.Core.Parser
+{
+    /// <summary>
+    /// Expression promoter is used whe matching
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/IExpressionPromoter.cs]: Typo and maybe add some more text to explain what this is doing.

File: src/System.Linq.Dynamic.Core/Parser/IExpressionPromoter.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    /// </summary>
+    public interface IExpressionPromoter
+    {
+        /// <summary>
+        /// Promote an expression
+        /// </summary>
+        /// <param name="expr">Source expression.</param>
+        /// <param name="type">Destionation data type to promote to.</param>
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/IExpressionPromoter.cs]: typo

File: src/System.Linq.Dynamic.Core/Parser/SupportedMethods/MethodFinder.cs [hunk 6]
```diff
@@ -4,14 +4,28 @@
...
 namespace System.Linq.Dynamic.Core.Parser.SupportedMethods
 {
-    internal static class MethodFinder
+    internal class MethodFinder
     {
-        public static bool ContainsMethod(Type type, string methodName, bool staticAccess, Expression[] args)
+        /// <summary>
+        /// The parsing config
+        /// </summary>
+        private ParsingConfig _parsingConfig;
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/SupportedMethods/MethodFinder.cs]: private readonly

File: src/System.Linq.Dynamic.Core/Parser/SupportedMethods/MethodFinder.cs [hunk 7]
```diff
@@ -4,14 +4,28 @@
...
+
+        /// <summary>
+        /// Get an instance
+        /// </summary>
+        /// <param name="parsingConfig"></param>
+        public MethodFinder(ParsingConfig parsingConfig)
+        {
+            this._parsingConfig = parsingConfig;
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/SupportedMethods/MethodFinder.cs]: `this.` not needed

File: src/System.Linq.Dynamic.Core/ParsingConfig.cs [hunk 8]
```diff
@@ -46,6 +55,25 @@ public IDynamicLinkCustomTypeProvider CustomTypeProvider
             }
         }
 
+        /// <summary>
+        /// Gets or sets the <see cref="IExpressionPromoter"/>.
+        /// </summary>
+        public IExpressionPromoter ExpressionPromoter
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/ParsingConfig.cs]: please add a unit test where you implement a different `ExpressionPromoter` or use `Mock<IExpressionPromoter>`

File: src/System.Linq.Dynamic.Core/ParsingConfig.cs [hunk 9]
```diff
@@ -20,8 +21,16 @@ public class ParsingConfig
             EvaluateGroupByAtDatabase = true
         };
 
+        /// <summary>
+        /// The custom type provider.
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/ParsingConfig.cs]: comment is not required for private

File: src/System.Linq.Dynamic.Core/ParsingConfig.cs [hunk 10]
```diff
@@ -20,8 +21,16 @@ public class ParsingConfig
...
+        /// </summary>
         private IDynamicLinkCustomTypeProvider _customTypeProvider;
 
+        /// <summary>
+        /// The expression promoter.
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/ParsingConfig.cs]: comment is not required for private

File: src/System.Linq.Dynamic.Core/Parser/SupportedMethods/MethodFinder.cs [hunk 11]
```diff
@@ -4,14 +4,28 @@
 
 namespace System.Linq.Dynamic.Core.Parser.SupportedMethods
 {
-    internal static class MethodFinder
+    internal class MethodFinder
     {
-        public static bool ContainsMethod(Type type, string methodName, bool staticAccess, Expression[] args)
+        /// <summary>
+        /// The parsing config
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/SupportedMethods/MethodFinder.cs]: comment is not required for private

**Comment by StefH** [comment]
@david-garcia-garcia Can you take a look at my review comments?

**Comment by david-garcia-garcia** [comment]
Took care of all comments + added some very basic test coverage.

**Pull request merged**